Added timer and stats for fps calculation

lancelot2112 opened the pull request:
Added timer and stats for throughput calculation to get an idea of where the bottlenecks are.

lancelot2112 commented:
![sim_time](https://user-images.githubusercontent.com/2574163/144719053-8b37abef-7154-494c-83c7-c381584d6c12.png)

Created some helper classes for timing.  not sure on best way to show stats other than console.log.  didn't want to hog throughput on that so updates every second.

MaxRobinsonTheGreat commented:
I'm actually seeing the infinite fps bug #42 in this. Also could you name your variables more descriptively? `val` and `take` in particular are confusing to me. What average is being taken with `take` anyway, average fps? time per tick?

lancelot2112 commented:
I'll have to look a little closer at the infinity, might just be that two calls to the function happened so close together that the timer resolution wasn't big enough to update?  I'll look closer.

It's a generic class, can be used anywhere you need a time or frequency of exec  It's not specific until you create an instance of it.  There are two instances one for the UI timer and one for the Simulation timer. 

Usage

sim_timer = new Timer();  //instance 1
ui_timer = new Timer();  //instance 2

   sim_update() {
       //takes internal state of the sim_timer {instance 1} and get delta time since last update call...  does not modify ui_timer
       sim_timer.update();  
       
       //does not change internal state of timer but calculates delta time since last update call and returns.  
       let sim_exec_time = sim_timer.tick();  
   }

the Stats class is also generic.  so it could be used to hold an average for anything.  Am I misunderstanding you perhaps?  So in the code there are two instances.  One for the "tick" stats and one for the "update" stats.

lancelot2112 commented:
Were you running firefox perchance?  Seems firefox does some rounding to the nearest millisecond, will protect for this.

https://stackoverflow.com/questions/6875625/does-javascript-provide-a-high-resolution-timer

lancelot2112 commented:
Changes to Timer class: changed "tick" to "time_since_last_update" and "update" to "update_state".  
Changes to Stats class: changed "take" to "add_value"

xkortex commented on src/Engine.js:
```diff
@@ -7,6 +7,62 @@ const ColorScheme = require('./Rendering/ColorScheme');
+        this.time = performance.now();
+
+        if(this.time == this.last_time) {
+            this.time = this.last_time + 0.5;
```
huh? What is `0.5` here? Are you just trying to avoid divide by zero? I don't think this is the best way to do that. 

If update state is called and the time resolution is too coarse to detect a change, then I think we need a different mechanism to determine FPS.

lancelot2112 replied:
Basically the issue is that our timers (on firefox in particular)  dont have sub millisecond resolution.  If the .now() function returns the same number in two subsequent calls then we are somewhere under or around 1ms though we dont know where.

In some cases its even worse from what Im reading, dont have sub 15ms resolution.  Im not sure best way to handle that.

lancelot2112 replied:
@xkortex @MaxRobinsonTheGreat 
https://humanwhocodes.com/blog/2011/12/14/timer-resolution-in-browsers/... looks like maybe 4ms is the minimum timer resolution?  Also it differs depending on browser.  4ms resolution would be a framerate of 250fps which would be why I see that number most of the time when it's smoothed out on Chromium Linux (which is what I'm running).  

Some other min resolution times i'm seeing that apparently depend on browser (eg. firefox) and operating system (eg. windows) which will affect the max framerate possible.  
4ms (250fps), 10ms (100fps), 15.6ms (~64fps)

I don't see an API to query this value though I think we could perhaps set up a "setTimeout" test where we set the timeout to 1ms and keep doing that until we see a time diff and then assign that time to a "min resolution" parameter?  

This would impact #62 perhaps.

lancelot2112 replied:
In latest commit added some "discovery" functionality that will look at the minimum timer resolution and find the minimum interrupt.  Changed this to just return if delta time is 0 to avoid div by zero.  Could update to assume the min resolution if you think that'd be prudent.
Running Chromium on Manjaro linux
![image](https://user-images.githubusercontent.com/2574163/145448622-fabbb664-9c00-4aa5-96ed-e62c8a75ed9c.png)

lancelot2112 replied:
Been thinking about ways to get this to run prior to LifeEngine running to avoid exec time of life engine getting in the way of the discovery process.  Haven't figured that part out yet.

lancelot2112 replied:
Moved exec to index.html.  
Firefox stats
![image](https://user-images.githubusercontent.com/2574163/145455497-abf5841a-efa2-4ee2-867d-e457be5fad36.png)

xkortex commented on src/Engine.js:
```diff
@@ -7,6 +7,62 @@ const ColorScheme = require('./Rendering/ColorScheme');
+        //takes a running average by weighting the current average as 75% of the 
+        //new value and the new value as 25%.  Adds a smoothing effect and works
+        //as a first order filter
+        this.average = (this.average * 3 + value)/4;
```
I would do this like 
```
this.average = this.weight * value  + (1 - this.weight * this.average)
```
and define `this.weight` in the constructor so you can easily adjust the smoothing function. 
see: https://en.wikipedia.org/wiki/Moving_average#:~:text=the%20ema%20for%20a%20series%20y%20%7B%5Cdisplaystyle%20y%7D%20may%20be%20calculated%20recursively%3A%20

lancelot2112 replied:
Good suggestion, will get that done
